### PR TITLE
[SPARK-5746][SQL] Check invalid cases for the write path of data source API

### PR DESCRIPTION
**Pull request by yhuai**
JIRA: https://issues.apache.org/jira/browse/SPARK-5746

@liancheng @marmbrus

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 1]
```diff
@@ -66,14 +66,17 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) {
   override protected[sql] def executePlan(plan: LogicalPlan): this.QueryExecution =
     new this.QueryExecution(plan)
 
+  @transient
+  protected[sql] val ddlParserWithHiveQL = new DDLParser(HiveQl.parseSql(_))
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: I am adding another ddl parser which takes `HiveQl.parseSql` to handle CTAS statements.

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala [hunk 2]
```diff
@@ -37,20 +37,22 @@ class ParquetQuerySuite extends QueryTest with ParquetTest {
...
+        sql("INSERT INTO TABLE t SELECT * FROM tmp")
         checkAnswer(table("t"), (data ++ data).map(Row.fromTuple))
       }
+      catalog.unregisterTable(Seq("tmp"))
     }
 
-    // This test case will trigger the NPE mentioned in
-    // https://issues.apache.org/jira/browse/PARQUET-151.
-    ignore(s"$prefix: overwriting") {
+    test(s"$prefix: overwriting") {
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala]: @liancheng I re-enabled this one. Seems it is fine now.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/rules.scala [hunk 3]
```diff
@@ -74,3 +74,68 @@ private[sql] object PreInsertCastAndRename extends Rule[LogicalPlan] {
...
+          failAnalysis(s"Insert into a partition is not allowed because $l is not partitioned.")
+        } else {
+          // Get all input data source relations of the query.
+          val srcRelations = query.collect {
+            case LogicalRelation(src: BaseRelation) => src
+          }
+          if (srcRelations.exists(src => src == t)) {
+            failAnalysis(
+              s"It is not allowed to insert overwrite the table because it is used as " +
+              s"an input table.")
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/rules.scala]: `Cannot insert overwrite into table that is also being read from`

File: sql/core/src/main/scala/org/apache/spark/sql/sources/rules.scala [hunk 4]
```diff
@@ -74,3 +74,68 @@ private[sql] object PreInsertCastAndRename extends Rule[LogicalPlan] {
...
+          } else {
+            // OK
+          }
+        }
+
+      case i @ logical.InsertIntoTable(l: LogicalRelation, partition, query, overwrite) =>
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/rules.scala]: Make the ! instanceOf here explicit.

**Pull request closed without merging**